flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStore.java

```diff
@@ -19,92 +19,224 @@
+	 * @param localState the local state to store.
+	 */
 	public void storeLocalState(
-		@Nonnull CheckpointMetaData checkpointMetaData,
+		@Nonnegative long checkpointId,
 		@Nullable TaskStateSnapshot localState) {
 
-		TaskStateSnapshot previous =
-			storedTaskStateByCheckpointID.put(checkpointMetaData.getCheckpointId(), localState);
+		if (localState == null) {
+			localState = NULL_DUMMY;
+		}
 
-		if (previous != null) {
-			throw new IllegalStateException("Found previously registered local state for checkpoint with id " +
-				checkpointMetaData.getCheckpointId() + "! This indicated a problem.");
+		LOG.info("Storing local state for checkpoint {}.", checkpointId);
+		LOG.debug("Local state for checkpoint {} is {}.", checkpointId, localState);
+
+		synchronized (lock) {
+			if (discarded) {
+				// we ignore late stores and simply discard the state.
+				discardStateObject(localState, checkpointId);
+			} else {
+				TaskStateSnapshot previous =
+					storedTaskStateByCheckpointID.put(checkpointId, localState);
+
+				if (previous != null) {
+					// this should never happen.
+					discardStateObject(previous, checkpointId);
+					throw new IllegalStateException("Found previously registered local state for checkpoint with id " +
+						checkpointId + "! This indicated a problem.");
+				}
+
+				// prune history.
+				while (storedTaskStateByCheckpointID.size() > MAX_RETAINED_SNAPSHOTS) {
+					Long removeCheckpointID = storedTaskStateByCheckpointID.firstKey();
+					TaskStateSnapshot snapshot =
+						storedTaskStateByCheckpointID.remove(storedTaskStateByCheckpointID.firstKey());
+					discardStateObject(snapshot, removeCheckpointID);
+				}
+			}
 		}
 	}
 
-	public void dispose() throws Exception {
-
-		Exception collectedException = null;
-
-		for (TaskStateSnapshot snapshot : storedTaskStateByCheckpointID.values()) {
-			try {
-				snapshot.discardState();
-			} catch (Exception discardEx) {
-				collectedException = ExceptionUtils.firstOrSuppressed(discardEx, collectedException);
-			}
+	/**
+	 * Returns the local state that is stored under the given checkpoint id or null if nothing was stored under the id.
+	 *
+	 * @param checkpointID the checkpoint id by which we search for local state.
+	 * @return the local state found for the given checkpoint id. Can be null
+	 */
+	@Nullable
+	public TaskStateSnapshot retrieveLocalState(long checkpointID) {
+		synchronized (lock) {
+			TaskStateSnapshot snapshot = storedTaskStateByCheckpointID.get(checkpointID);
+			return snapshot != NULL_DUMMY ? snapshot : null;
 		}
+	}
 
-		if (collectedException != null) {
-			throw collectedException;
+	/**
+	 * Disposes the state of all local snapshots managed by this object.
+	 */
+	public void dispose() {
+		synchronized (lock) {
+			for (Map.Entry<Long, TaskStateSnapshot> entry : storedTaskStateByCheckpointID.entrySet()) {
+				discardStateObject(entry.getValue(), entry.getKey());
+			}
+			discarded = true;
```

Comment:
Shall we rename `discarded` into `disposed`?

Reply:
👍